PXB-8.0-2740 PXB displays the encryption key when the backup is decrypted…

rahulmalik87 opened the pull request:
… and decompressed

https://jira.percona.com/browse/PXB-2740

Problem:
PXB displays the encryption key when the backup is decrypted and decompressed

Analysis:
PXB uses system function to call xbcrypt with encryption password to decrypt
the password. system function creates sub process. password is only
removed by sub process . The main process shows the encrypted password
in ps

Fix:
set the enviorment variable XBCRYPT_ENCRYPTION_PASSWORD in pxb and use
that variable inside xbcrypt

satya-bodapati commented on storage/innobase/xtrabackup/src/xbcrypt_common.cc:
```diff
@@ -305,3 +305,10 @@ gcry_error_t xb_crypt_encrypt(gcry_cipher_hd_t cipher_handle, const uchar *from,
+void get_env_value(char *&var, const char *env) {
+  char *val;
+  if (var == nullptr && (val = getenv(env)) != nullptr) {
+    var = my_strdup(PSI_NOT_INSTRUMENTED, val, MYF(MY_FAE));
```
where do we free this?

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -2584,9 +2585,7 @@ bool decrypt_decompress_file(const char *filepath, uint thread_n) {
-    if (xtrabackup_encrypt_key) {
-      cmd << " --encrypt-key=" << xtrabackup_encrypt_key;
-    } else {
+    if (xtrabackup_encrypt_key == nullptr) {
```
can we do setenv() here?
if (xtrabckup_encrypt_key) {
setenv();
} else {
--file
}

rahulmalik87 replied:
@satya-bodapati , if we do setenv inside decrypt_decompress_file we have to do it for each file. better to do it outside of this function so global environment will be called once .

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -2584,9 +2585,7 @@ bool decrypt_decompress_file(const char *filepath, uint thread_n) {
-    if (xtrabackup_encrypt_key) {
-      cmd << " --encrypt-key=" << xtrabackup_encrypt_key;
-    } else {
+    if (xtrabackup_encrypt_key == nullptr) {
```
also, I think it is better to unset it after usage.
get old value (getnv()), save it. and restore it after system() call.

satya-bodapati replied:
on second thought, I think, the ENV change done by this xtrabackup process will not be visible to other xtrabackup processes. So pxb exit will auto unset it.

Please verify

rahulmalik87 replied:
correct, pxb exists with auto unset.

satya-bodapati reviewed (changes requested):
also better to do GCA and make PR for 2.4

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -2655,6 +2654,10 @@ static void decrypt_decompress_thread_func(datadir_thread_ctxt_t *ctxt) {
       continue;
     }
 
+    if (xtrabackup_encrypt_key) {
+      setenv("XBCRYPT_ENCRYPTION_KEY", xtrabackup_encrypt_key, 1);
+    }
+
```
here too, we seem to set after every file, this should be set before the while loop isn't it?

Commit:
PXB-2740 PXB displays the encryption key when the backup is decrypted and decompressed
https://jira.percona.com/browse/PXB-2740

Problem:
PXB displays the encryption key when the backup is decrypted and decompressed

Analysis:
PXB uses system function to call xbcrypt with encryption password to decrypt
the password. system function creates sub process. password is only
removed by sub process . The main process shows the encrypted password
in ps

Fix:
set the enviorment variable XBCRYPT_ENCRYPTION_KEY in pxb and use
that variable inside xbcrypt

Commit:
Merge branch 'pxb-2.4-2740' into pxb-8.0-2740

rahulmalik87 commented:
@satya-bodapati  test results.



==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
check_for_passwords                      w1	[passed]    60
==============================================================================
Spent 60 of 86 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed
